### PR TITLE
ArtifactStore implementation for MongoDB

### PR DESCRIPTION
**Pull request by jiangpengcheng**
This PR provides a MongoDBArtifactStore implementation to enable using MongoDB for storing subjects, whisks and activation in [MongoDB](https://www.mongodb.com/)

## Description
Some users may want to use other database backends instead of CouchDB to store entities, this PR provides MongoDB as an alternative choice for these users

## MongoDB driver
MongoDB provides [official drivers](https://docs.mongodb.com/ecosystem/drivers/) for many languages, in this case, we pick the [mongo-scala-driver](http://mongodb.github.io/mongo-scala-driver/2.2/), this driver provides an idiomatic Scala API that is built on top of the MongoDB Async Java driver.

### Data Model
The data scheme in MongoDB is just like CouchDB except that MongoDB doesn't have a `_rev` field, below is the data of `whisk.system/invokerHealthTestAction0` in CouchDB and MongoDB:

<table>
<tr>
<td>CouchDB</td>
<td>MongoDB</td>
</tr>
<tr>
<td>
<pre>
{
  "_id": "whisk.system/invokerHealthTestAction0",
  "_rev": "68-e72440f911c64ab11441c09e730e5ab8",
  "name": "invokerHealthTestAction0",
  "publish": false,
  "annotations": [],
  "version": "0.0.1",
  "updated": 1524476933182,
  "entityType": "action",
  "exec": {
    "kind": "nodejs:6",
    "code": "function main(params) { return params; }",
    "binary": false
  },
  "parameters": [],
  "limits": {
    "timeout": 60000,
    "memory": 256,
    "logs": 10
  },
  "namespace": "whisk.system"
}
</pre>
</td>
<td>
<pre>
{
  "_id" : "whisk.system/invokerHealthTestAction0",
  "name" : "invokerHealthTestAction0",
  "publish" : false,
  "annotations" : [ ],
  "version" : "0.0.1",
  "updated" : NumberLong("1524473794826"),
  "entityType" : "action",
  "exec" : {
    "kind" : "nodejs:6",
    "code" : "function main(params) { return params; }",
    "binary" : false
  },
  "parameters" : [ ],
  "limits" : {
    "timeout" : 60000,
    "memory" : 256,
    "logs" : 10
  },
  "namespace" : "whisk.system"
}
</pre>
</td>
</tr>
</table>

*The MongoDB use BSON to store data,  `MongoDbStrore` will convert bson to json, this is transparent to users*

Regarding the `_rev` field, there is no alternative way in MongoDB to do the same thing, so the `MongoDbStore` calculate a value according to the document's content manually, and make it as the `_rev` field.

### Attachment
MongoDB use [GridFS](https://docs.mongodb.com/manual/core/gridfs/index.html) to store and retrieve files that exceed the BSON-document size limit of 16 MB.

Attachment in MongoDB is stored in a separate collection with a independent `_id`, this PR use the `doc._id + doc.file_name` as the attachment's `_id` field, then we can find the relative attachment easily.

### Query
CouchDB use `designDoc` to do the query operation, ~~MongoDB can use `view` feature to achieve the same effect, there are some examples below~~:

#### whisks-filters.v2.1.0/activations
```
db.createView(
    "whisk_local_whisks.whisks-filters.v2.1.0/activations",
    "whisk_local_whisks",
    [
        { "$match": { "activationId": { "$exists": true } } },
        {
            "$addFields": {
                "flag": {
                    "$arrayElemAt": [{ "$filter": { "input": "$annotations", "cond": { "$eq": ["$$this.key", "path"] } } }, 0]
                }
            }
        },
        { "$lookup": { "from": "whisk_local_whisks", "localField": "_id", "foreignField": "_id", "as": "doc" } },
        {
            "$project": {
                "key": [{ "$ifNull": ["$flag.value", { "$concat": ["$namespace", "/", "$name"] }] }, "$start"],
                "value": {
                    "namespace": "$namespace",
                    "name": "$name",
                    "version": "$version",
                    "publish": "$publish",
                    "annotations": "$annotations",
                    "activationId": "$activationId",
                    "start": "$start",
                    "end": "$end",
                    "duration": "$duration",
                    "cause": "$cause",
                    "statusCode": "$response.statusCode"
                },
                "_id": 0,
                "id": "$_id",
                "doc": { "$arrayElemAt": ["$doc", 0] }
            }
        }
    ]
)
```

As @chetanmeh suggest and I test again with 300,000 activations, the performance of `view` in MongoDB is not good as native query while call `wsk activation list`, so I change strategies and follow the [openwhisk-mongo](https://github.com/chetanmeh/openwhisk-mongo)

### Reduce Design Doc
~~Unfortunately, there is no easy way to do the same thing with the CouchDB style `reduce` in MongoDB, because of `reduce` in CouchDB can use `start_key` 、`end_key` as filters to produce a different result while `reduce` in MongoDB can not, so I do a trade-off to simulate some simple CouchDB style `reduce`~~, currently, this pr doesn't support the `Reduce` feature, need investigate it later

## MongoDB Deployment
Currently only support deploying MongoDB replica set.

### Finished Work
- [x] Basic Usage(create、update、query、delete)
- [x] Attachment Support
- [x] Automated deployment of MongoDB using Ansible
- [x] CI Integration

- [x] Documents

### FurtherWork
- [ ] Add MongoDB support for wskadmin

## My changes affect the following components
- [x] Data stores (e.g., CouchDB)
- [x] Tests
- [x] Deployment
- [x] CLI
- [x] Documentation

## Types of changes
- [x] Enhancement or new feature (adds new functionality).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [x] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [x] I updated the documentation where necessary.

### CODE REVIEW

**Comment by junoyoon** [comment]
rebase please.

**Comment by junoyoon** [comment]
rebase please.

**Comment by jiangpengcheng** [comment]
@chetanmeh 
thanks for your review! I will check it

**Comment by jiangpengcheng** [comment]
@chetanmeh, I followed many idead from your project https://github.com/chetanmeh/openwhisk-mongo, so I added you as the co-author

@dubee @chetanmeh  please re-review this

**Comment by chetanmeh** [comment]
@jiangpengcheng Thanks for updating the PR to use query. I would try to review it this week. It may take some time

File: tests/src/test/resources/application.conf.j2 [hunk 1]
```diff
@@ -55,6 +55,21 @@ whisk {
...
     }
 
+    {% if enable_mongodb is defined and enable_mongodb == 'true' %}
+    mongodb {
+        host        = "{{ db.host }}"
+        port        = "{{ mongodb_port| default(db.port) }}"
+        database    = "{{ db.mongodb.database }}"
+        username    = "{{ db.credentials.admin.user }}"
+        password    = "{{ db.credentials.admin.pass }}"
+        collections {
```
[the remark this answers is not in the document]
Comment (jiangpengcheng) [reply tests/src/test/resources/application.conf.j2]: I'm sorry, but I didn't get you meaning, do you mean not set the `collections` option in the configuration?

File: common/scala/src/main/scala/whisk/core/database/mongodb/MongoDBArtifactStore.scala [hunk 2]
```diff
@@ -0,0 +1,622 @@
...
+
+import scala.concurrent.Future
+import scala.util.Try
+
+object MongoDBArtifactStore {
+  val _computed = "_computed"
+  implicit class StringPurge(val s: String) {
+    def escapeDollar: String = {
+      // replace '"$xxx":' to '"_mark_of_dollar_xxx":' as the "$" can not be the first char of field name
+      s.replaceAll("\"\\$([^}]+?\":)", "\"_mark_of_dollar_$1")
```
[the remark this answers is not in the document]
Comment (jiangpengcheng) [reply common/scala/src/main/scala/whisk/core/database/mongodb/MongoDBArtifactStore.scala]: yes, there are some examples in the openwhisk-catalog, e.g. https://github.com/apache/incubator-openwhisk-catalog/blob/master/packages/installCombinators.sh#L51, is it ok to ignore them?

File: common/scala/build.gradle [hunk 3]
```diff
@@ -87,12 +87,16 @@ dependencies {
     compile 'io.reactivex:rxjava-reactive-streams:1.2.1'
     compile 'com.microsoft.azure:azure-cosmosdb:2.3.0'
 
+    //for mongo
+    compile 'org.mongodb.scala:mongo-scala-driver_2.12:2.6.0'
```
Comment (rabbah) [inline common/scala/build.gradle]: i rebased this pr to master, and resolved conflicts.
in doing this, i picked up the latest version of the driver which might have been foolish.
please check it.

**Comment by jiangpengcheng** [comment]
closed as this is outdated

**Comment by LendlMedina** [comment]
Hi,

I was running into an error with 

cd openwhisk-mongo
./gradlew distDocker 

Could not resolve all files for configuration ':common:mongo:compileClasspath'.
> Could not find org.apache.openwhisk:openwhisk-common:1.0.0-SNAPSHOT.
Searche in the following locations:
https://repo.maven.apache.org/maven2/org/apache/openwhisk/openwhisk-common/1.0.0-SNAPSHOT/maven-metadata.xml
https://repo.maven.apache.org/maven2/org/apache/openwhisk/openwhisk-common/1.0.0-SNAPSHOT/openwhisk-common-1.0.0-SNAPSHOT.pom
https://repo.maven.apache.org/maven2/org/apache/openwhisk/openwhisk-common/1.0.0-SNAPSHOT/openwhisk-common-1.0.0-SNAPSHOT.jar
Required by:
project :common:mongo


Im running with
uname -a
Linux ip-10-10-110-212 4.15.0-1051-aws #53-Ubuntu SMP Wed Sep 18 13:35:53 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

java -version
openjdk version "1.8.0_222"
OpenJDK Runtime Environment (build 1.8.0_222-8u222-b10-1ubuntu1~18.04.1-b10)
OpenJDK 64-Bit Server VM (build 25.222-b10, mixed mode)

docker-compose --version
docker-compose version 1.17.0, build ac53b73

**Pull request closed without merging**